[Quant] lower fused LinearTanh for onednn backend

Xia-Weiwen opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #89188
* #88923
* #88879
* #88668


**Summary**
Add fuser method and quantization mappings for `QLinearLeakyReLU` for int8 inference for onednn backend. The fusion and lowering are supported only in FX mode.

**Test plan**
python test_quantization.py TestFuseFx TestQuantizeFx


cc @jerryzh168 @jianyuh @raghuramank100 @jamesr66a @vkuzo @jgong5 @leslie-fang-intel @mingfeima @XiaobingSuper @sanchitintel @ashokei @jingxu10

Xia-Weiwen committed:
[Quant] lower fused LinearTanh for onednn backend

[ghstack-poisoned]

Xia-Weiwen committed:
Update on "[Quant] lower fused LinearTanh for onednn backend"




cc jerryzh168 jianyuh raghuramank100 jamesr66a vkuzo jgong5 leslie-fang-intel

[ghstack-poisoned]

Xia-Weiwen committed:
Update on "[Quant] lower fused LinearTanh for onednn backend"




cc jerryzh168 jianyuh raghuramank100 jamesr66a vkuzo jgong5 leslie-fang-intel

[ghstack-poisoned]

Xia-Weiwen committed:
Update on "[Quant] lower fused LinearTanh for onednn backend"




cc jerryzh168 jianyuh raghuramank100 jamesr66a vkuzo jgong5 leslie-fang-intel

[ghstack-poisoned]

Xia-Weiwen committed:
Update on "[Quant] lower fused LinearTanh for onednn backend"




cc jerryzh168 jianyuh raghuramank100 jamesr66a vkuzo jgong5 leslie-fang-intel

[ghstack-poisoned]

Xia-Weiwen committed:
Update on "[Quant] lower fused LinearTanh for onednn backend"

**Summary**
Add fuser method and quantization mappings for `QLinearLeakyReLU` for int8 inference for onednn backend. The fusion and lowering are supported only in FX mode.

**Test plan**
python test_quantization.py TestFuseFx TestQuantizeFx




cc jerryzh168 jianyuh raghuramank100 jamesr66a vkuzo jgong5 leslie-fang-intel

[ghstack-poisoned]

Xia-Weiwen committed:
Update on "[Quant] lower fused LinearTanh for onednn backend"



**Summary**
Add fuser method and quantization mappings for `QLinearLeakyReLU` for int8 inference for onednn backend. The fusion and lowering are supported only in FX mode.

**Test plan**
python test_quantization.py TestFuseFx TestQuantizeFx


cc jerryzh168 jianyuh raghuramank100 jamesr66a vkuzo jgong5 leslie-fang-intel mingfeima XiaobingSuper sanchitintel ashokei jingxu10

[ghstack-poisoned]

Xia-Weiwen committed:
Update on "[Quant] lower fused LinearTanh for onednn backend"



**Summary**
Add fuser method and quantization mappings for `QLinearLeakyReLU` for int8 inference for onednn backend. The fusion and lowering are supported only in FX mode.

**Test plan**
python test_quantization.py TestFuseFx TestQuantizeFx


cc jerryzh168 jianyuh raghuramank100 jamesr66a vkuzo jgong5 leslie-fang-intel mingfeima XiaobingSuper sanchitintel ashokei jingxu10

[ghstack-poisoned]

Xia-Weiwen committed:
Update on "[Quant] lower fused LinearTanh for onednn backend"



**Summary**
Add fuser method and quantization mappings for `QLinearLeakyReLU` for int8 inference for onednn backend. The fusion and lowering are supported only in FX mode.

**Test plan**
python test_quantization.py TestFuseFx TestQuantizeFx


cc jerryzh168 jianyuh raghuramank100 jamesr66a vkuzo jgong5 leslie-fang-intel mingfeima XiaobingSuper sanchitintel ashokei jingxu10

[ghstack-poisoned]

Xia-Weiwen committed:
Update on "[Quant] lower fused LinearTanh for onednn backend"



**Summary**
Add fuser method and quantization mappings for `QLinearLeakyReLU` for int8 inference for onednn backend. The fusion and lowering are supported only in FX mode.

**Test plan**
python test_quantization.py TestFuseFx TestQuantizeFx


cc jerryzh168 jianyuh raghuramank100 jamesr66a vkuzo jgong5 leslie-fang-intel mingfeima XiaobingSuper sanchitintel ashokei jingxu10

[ghstack-poisoned]

Xia-Weiwen committed:
Update on "[Quant] lower fused LinearTanh for onednn backend"



**Summary**
Add fuser method and quantization mappings for `QLinearLeakyReLU` for int8 inference for onednn backend. The fusion and lowering are supported only in FX mode.

**Test plan**
python test_quantization.py TestFuseFx TestQuantizeFx


cc jerryzh168 jianyuh raghuramank100 jamesr66a vkuzo jgong5 leslie-fang-intel mingfeima XiaobingSuper sanchitintel ashokei jingxu10

[ghstack-poisoned]

Xia-Weiwen committed:
Update on "[Quant] lower fused LinearTanh for onednn backend"



**Summary**
Add fuser method and quantization mappings for `QLinearLeakyReLU` for int8 inference for onednn backend. The fusion and lowering are supported only in FX mode.

**Test plan**
python test_quantization.py TestFuseFx TestQuantizeFx


cc jerryzh168 jianyuh raghuramank100 jamesr66a vkuzo jgong5 leslie-fang-intel mingfeima XiaobingSuper sanchitintel ashokei jingxu10

[ghstack-poisoned]

Xia-Weiwen commented:
Hi @jgong5 Could you review? Thanks!

Xia-Weiwen committed:
Update on "[Quant] lower fused LinearTanh for onednn backend"



**Summary**
Add fuser method and quantization mappings for `QLinearLeakyReLU` for int8 inference for onednn backend. The fusion and lowering are supported only in FX mode.

**Test plan**
python test_quantization.py TestFuseFx TestQuantizeFx


cc jerryzh168 jianyuh raghuramank100 jamesr66a vkuzo jgong5 leslie-fang-intel mingfeima XiaobingSuper sanchitintel ashokei jingxu10

[ghstack-poisoned]

Xia-Weiwen commented:
Hi @jerryzh168 Could you help review this PR please? Thanks!

Xia-Weiwen committed:
Update on "[Quant] lower fused LinearTanh for onednn backend"



**Summary**
Add fuser method and quantization mappings for `QLinearLeakyReLU` for int8 inference for onednn backend. The fusion and lowering are supported only in FX mode.

**Test plan**
python test_quantization.py TestFuseFx TestQuantizeFx


cc jerryzh168 jianyuh raghuramank100 jamesr66a vkuzo jgong5 leslie-fang-intel mingfeima XiaobingSuper sanchitintel ashokei jingxu10

[ghstack-poisoned]

Xia-Weiwen committed:
Update on "[Quant] lower fused LinearTanh for onednn backend"



**Summary**
Add fuser method and quantization mappings for `QLinearLeakyReLU` for int8 inference for onednn backend. The fusion and lowering are supported only in FX mode.

**Test plan**
python test_quantization.py TestFuseFx TestQuantizeFx


cc jerryzh168 jianyuh raghuramank100 jamesr66a vkuzo jgong5 leslie-fang-intel mingfeima XiaobingSuper sanchitintel ashokei jingxu10

[ghstack-poisoned]

Xia-Weiwen committed:
Update on "[Quant] lower fused LinearTanh for onednn backend"



**Summary**
Add fuser method and quantization mappings for `QLinearLeakyReLU` for int8 inference for onednn backend. The fusion and lowering are supported only in FX mode.

**Test plan**
python test_quantization.py TestFuseFx TestQuantizeFx


cc jerryzh168 jianyuh raghuramank100 jamesr66a vkuzo jgong5 leslie-fang-intel mingfeima XiaobingSuper sanchitintel ashokei jingxu10

[ghstack-poisoned]

Xia-Weiwen committed:
Update on "[Quant] lower fused LinearTanh for onednn backend"



**Summary**
Add fuser method and quantization mappings for `QLinearLeakyReLU` for int8 inference for onednn backend. The fusion and lowering are supported only in FX mode.

**Test plan**
python test_quantization.py TestFuseFx TestQuantizeFx


cc jerryzh168 jianyuh raghuramank100 jamesr66a vkuzo jgong5 leslie-fang-intel mingfeima XiaobingSuper sanchitintel ashokei jingxu10

[ghstack-poisoned]

jerryzh168 commented on test/quantization/fx/test_quantize_fx.py:
```diff
@@ -5619,39 +5660,68 @@ def forward(self, x):
-        """
-        from torch.ao.quantization.backend_config import get_onednn_backend_config
-        qconfig_mapping = get_default_qconfig_mapping('onednn')
+    def _test_activation_fusion_lowering_helper(
```
nit: maybe `_test_linear_activation_fusion_lowering_helper`

Xia-Weiwen replied:
Ok. It's fixed.

Xia-Weiwen committed:
Update on "[Quant] lower fused LinearTanh for onednn backend"



**Summary**
Add fuser method and quantization mappings for `QLinearLeakyReLU` for int8 inference for onednn backend. The fusion and lowering are supported only in FX mode.

**Test plan**
python test_quantization.py TestFuseFx TestQuantizeFx


cc jerryzh168 jianyuh raghuramank100 jamesr66a vkuzo jgong5 leslie-fang-intel mingfeima XiaobingSuper sanchitintel ashokei jingxu10

[ghstack-poisoned]

jerryzh168 commented on torch/ao/quantization/backend_config/onednn.py:
```diff
@@ -119,47 +119,58 @@ def _fuse_linear_bn_leaky_relu(is_qat, linear, bn, leaky_relu):
+                                ref_quant_module):
+    # 1 base module + op module fusion config
+    configs.append(
+        BackendPatternConfig((post_module, root_module))
```
nit: the pattern need to be updated since we recently landed: https://github.com/pytorch/pytorch/pull/90698

Xia-Weiwen replied:
Ok. It's fixed.

Xia-Weiwen committed:
Update on "[Quant] lower fused LinearTanh for onednn backend"



**Summary**
Add fuser method and quantization mappings for `QLinearLeakyReLU` for int8 inference for onednn backend. The fusion and lowering are supported only in FX mode.

**Test plan**
python test_quantization.py TestFuseFx TestQuantizeFx


cc jerryzh168 jianyuh raghuramank100 jamesr66a vkuzo jgong5 leslie-fang-intel mingfeima XiaobingSuper sanchitintel ashokei jingxu10

[ghstack-poisoned]

Xia-Weiwen committed:
Update on "[Quant] lower fused LinearTanh for onednn backend"



**Summary**
Add fuser method and quantization mappings for `QLinearLeakyReLU` for int8 inference for onednn backend. The fusion and lowering are supported only in FX mode.

**Test plan**
python test_quantization.py TestFuseFx TestQuantizeFx


cc jerryzh168 jianyuh raghuramank100 jamesr66a vkuzo jgong5 leslie-fang-intel mingfeima XiaobingSuper sanchitintel ashokei jingxu10

[ghstack-poisoned]

Xia-Weiwen committed:
Update on "[Quant] lower fused LinearTanh for onednn backend"



**Summary**
Add fuser method and quantization mappings for `QLinearLeakyReLU` for int8 inference for onednn backend. The fusion and lowering are supported only in FX mode.

**Test plan**
python test_quantization.py TestFuseFx TestQuantizeFx


cc jerryzh168 jianyuh raghuramank100 jamesr66a vkuzo jgong5 leslie-fang-intel mingfeima XiaobingSuper sanchitintel ashokei jingxu10

[ghstack-poisoned]

Xia-Weiwen committed:
Update on "[Quant] lower fused LinearTanh for onednn backend"



**Summary**
Add fuser method and quantization mappings for `QLinearLeakyReLU` for int8 inference for onednn backend. The fusion and lowering are supported only in FX mode.

**Test plan**
python test_quantization.py TestFuseFx TestQuantizeFx


cc jerryzh168 jianyuh raghuramank100 jamesr66a vkuzo jgong5 leslie-fang-intel mingfeima XiaobingSuper sanchitintel ashokei jingxu10

[ghstack-poisoned]

Xia-Weiwen committed:
Update on "[Quant] lower fused LinearTanh for onednn backend"



**Summary**
Add fuser method and quantization mappings for `QLinearLeakyReLU` for int8 inference for onednn backend. The fusion and lowering are supported only in FX mode.

**Test plan**
python test_quantization.py TestFuseFx TestQuantizeFx


cc jerryzh168 jianyuh raghuramank100 jamesr66a vkuzo jgong5 leslie-fang-intel mingfeima XiaobingSuper sanchitintel ashokei jingxu10

[ghstack-poisoned]

Xia-Weiwen committed:
Update on "[Quant] lower fused LinearTanh for onednn backend"



**Summary**
Add fuser method and quantization mappings for `QLinearLeakyReLU` for int8 inference for onednn backend. The fusion and lowering are supported only in FX mode.

**Test plan**
python test_quantization.py TestFuseFx TestQuantizeFx


cc jerryzh168 jianyuh raghuramank100 jamesr66a vkuzo jgong5 leslie-fang-intel mingfeima XiaobingSuper sanchitintel ashokei jingxu10

[ghstack-poisoned]

Xia-Weiwen committed:
Update on "[Quant] lower fused LinearTanh for onednn backend"



**Summary**
Add fuser method and quantization mappings for `QLinearLeakyReLU` for int8 inference for onednn backend. The fusion and lowering are supported only in FX mode.

**Test plan**
python test_quantization.py TestFuseFx TestQuantizeFx


cc jerryzh168 jianyuh raghuramank100 jamesr66a vkuzo jgong5 leslie-fang-intel mingfeima XiaobingSuper sanchitintel ashokei jingxu10

[ghstack-poisoned]

Xia-Weiwen committed:
Update on "[Quant] lower fused LinearTanh for onednn backend"



**Summary**
Add fuser method and quantization mappings for `QLinearLeakyReLU` for int8 inference for onednn backend. The fusion and lowering are supported only in FX mode.

**Test plan**
python test_quantization.py TestFuseFx TestQuantizeFx


cc jerryzh168 jianyuh raghuramank100 jamesr66a vkuzo jgong5 leslie-fang-intel mingfeima XiaobingSuper sanchitintel ashokei jingxu10

[ghstack-poisoned]

Xia-Weiwen committed:
Update on "[Quant] lower fused LinearTanh for onednn backend"



**Summary**
Add fuser method and quantization mappings for `QLinearLeakyReLU` for int8 inference for onednn backend. The fusion and lowering are supported only in FX mode.

**Test plan**
python test_quantization.py TestFuseFx TestQuantizeFx


cc jerryzh168 jianyuh raghuramank100 jamesr66a vkuzo jgong5 leslie-fang-intel mingfeima XiaobingSuper sanchitintel ashokei jingxu10

[ghstack-poisoned]

Xia-Weiwen committed:
Update on "[Quant] lower fused LinearTanh for onednn backend"



**Summary**
Add fuser method and quantization mappings for `QLinearLeakyReLU` for int8 inference for onednn backend. The fusion and lowering are supported only in FX mode.

**Test plan**
python test_quantization.py TestFuseFx TestQuantizeFx


cc jerryzh168 jianyuh raghuramank100 jamesr66a vkuzo jgong5 leslie-fang-intel mingfeima XiaobingSuper sanchitintel ashokei jingxu10

[ghstack-poisoned]

Xia-Weiwen committed:
Update on "[Quant] lower fused LinearTanh for onednn backend"



**Summary**
Add fuser method and quantization mappings for `QLinearLeakyReLU` for int8 inference for onednn backend. The fusion and lowering are supported only in FX mode.

**Test plan**
python test_quantization.py TestFuseFx TestQuantizeFx


cc jerryzh168 jianyuh raghuramank100 jamesr66a vkuzo jgong5 leslie-fang-intel mingfeima XiaobingSuper sanchitintel ashokei jingxu10

[ghstack-poisoned]

Xia-Weiwen commented:
Hi @jerryzh168 I have made changes per your comments. Please take a look again. Thanks!

Xia-Weiwen commented:
@pytorchbot merge
